### PR TITLE
Fixed #6978: DI Container is not reset when destroying application in functional tests

### LINKED ISSUE #6978: Dependency Injection Container not reset when destroying application in functional tests
I use `\Yii::$container` to override some classes during unit tests. Functional tests run after unit tests and fail, because `\Yii::$container->get` still returns the temporary PHPUnit Mock objects rather than the real classes. 

Should not `\yii\codeception\TestCase::destroyApplication` also clear the Dependency Injection Container like this? 

```php
    /**
     * Destroys the application instance created by [[mockApplication]].
     */
    protected function destroyApplication()
    {
        Yii::$app = null;
        Yii::$container = new \yii\di\Container()
    }
```
I can provide a pull request if this is the correct way of solving this.

### PR DESCRIPTION
**Pull request by ivokund**
Added code to reset the DI container when destroying the application

### CODE REVIEW

**Comment by samdark** [comment]
@yiisoft/core-developers, @Ragazzo any issues with the change?

**Comment by klimov-paul** [comment]
I am unsure: it could be someone relies on DI for test environment setup.
It could be better to save original instance if `Container` before running the test and then restore it, rather then simply recreating it.

**Comment by samdark** [comment]
@klimov-paul environment should be re-created for each test case i.e. it should be isolated.

**Comment by Ragazzo** [comment]
i dont think that it will affect anything , since tests should be done in isolation , however we should recheck that DI - container is recreated each time in mockApplication()

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@ Yii Framework 2 Change Log
 
 - Bug #6919: Fixed wrong namespaces under advanced application's TestCase classes (ivokund)
 - Bug #6969: `yii\helpers\ArrayHelper::htmlEncode()` and `htmlDecode()` should not remove non-string data (qiangxue)
+- Bug #6978: DI Container is not reset when destroying application in functional tests (ivokund)
```
Comment (cebe) [inline framework/CHANGELOG.md]: changelog has to be in https://github.com/yiisoft/yii2/blob/master/extensions/codeception/CHANGELOG.md

**Comment by cebe** [comment]
> however we should recheck that DI - container is recreated each time in mockApplication()

:+1:

**Comment by ivokund** [comment]
Okay, do I understand this correctly now?
To be precise, we should do `Yii::$container = null` (to destroy is to set to default value) in `destroyApplication()` and  `Yii::$container = new Container` in `mockApplication()` (to mimick the first load of `Yii` class?)
```php
class Yii extends \yii\BaseYii
{
}

spl_autoload_register(['Yii', 'autoload'], true, true);
Yii::$classMap = include(__DIR__ . '/classes.php');
Yii::$container = new yii\di\Container;
```

**Comment by ivokund** [comment]
Ok, setting it to null is not a good idea, because `yii\base\Application::__construct` needs the container to be set already.

**Comment by ivokund** [comment]
Added the change to the correct changelog, added the reset of DI container also to `mockApplication`.

Did you mean something like this by "recheck"?

**Comment by samdark** [comment]
Merged. Thanks! https://github.com/yiisoft/yii2/commit/a6dd9d37d5c5b7e5fac4edeb89c35cbc961df369

**Pull request closed without merging**